From 589b7e9600b1e50574ce91e4d6fc903ef81d957f Mon Sep 17 00:00:00 2001
From: INAGAKI Hiroshi <musashino.open@gmail.com>
Date: Fri, 14 Feb 2025 18:19:58 +0900
Subject: [PATCH] spi: spi-qpic-snand: add user_config support for ECC

Make the ECC strength in the qpic-snand driver configurable via device
tree to use the device-specific ECC strength value.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>
---
--- a/drivers/spi/spi-qpic-snand.c
+++ b/drivers/spi/spi-qpic-snand.c
@@ -274,8 +274,20 @@ static int qcom_spi_ecc_init_ctx_pipelin
 	nand->ecc.ctx.priv = ecc_cfg;
 	snandc->qspi->mtd = mtd;
 
-	/* BCH8 or BCH4 */
-	ecc_mode = mtd->oobsize > 64 ? 1 : 0;
+	/* BCH4 or BCH8 */
+	switch (nand->ecc.user_conf.strength) {
+	case 4:
+		ecc_mode = 0;
+		break;
+	case 8:
+		ecc_mode = 1;
+		break;
+	default:
+		ecc_mode = mtd->oobsize > 64 ? 1 : 0;
+		dev_warn(snandc->dev,
+			 "invalid ECC strength detected, configured from mtd oobsize\n");
+		break;
+	}
 
 	ecc_cfg->ecc_bytes_hw = ecc_mode ? 13 : 7;
 	ecc_cfg->spare_bytes = ecc_mode ? 2 : 4;
